sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala

```diff
@@ -203,6 +203,21 @@ trait FileFormat {
    * method. Technically, a file format could choose suppress them, but that is not recommended.
    */
   def metadataSchemaFields: Seq[StructField] = FileFormat.BASE_METADATA_FIELDS
+
+  /**
+   * The extractors to use when deriving file-constant metadata columns for this file format.
+   *
+   * A scanner must derive each file-constant metadata field's value from each [[PartitionedFile]]
+   * it processes. By default, the value is obtained by a direct lookup of the column's name on
+   * [[PartitionedFile.otherConstantMetadataColumnValues]] (see
+   * [[FileFormat.getFileConstantMetadataColumnValue]]). However, implementations can override this
+   * method in order to provide more sophisticated lazy extractors (e.g. in case the column value is
+   * complicated or expensive to compute).
```

Comment:
Someone adding a new constant metadata column in the future will need to choose between the two alternatives to provide values, this choice could be described more explicitly here:

- extractor: anytime the value can be directly derived from `PartitionedFile` information.
- `PartitionedFile.otherConstantMetadataColumnValues`: all other cases, e.g. the value is arbitrary or is coming from some other place

Reply:
I thought I _did_ describe it explicitly:
1. If you provide an extractor, the extractor has access to all state in the `PartitionedFile` (including the column value map) and can do any computations it needs to.
2. Otherwise, the column's value is fetched from the column value map.

Was it not clear [enough] in the comment that the extractor has access to the entire `PartitionedFile` ?

Reply:
Reworded, PTAL

Reply:
The way I read it is that using `otherConstantMetadataColumnValues` is the default and an extractor should only be used for more advanced use cases. I think we rather want to encourage using extractors whenever possible and only fallback to manually providing the values when necessary.

Reply:
That wording was intentional... the simple way is usually the right way and thus should be the default. If something special is going on, I'd prefer it to be highly visible. 

Put another way -- if the column _can_ be represented simply by pulling the value from the map, why should we encourage the use of an extractor? 

An alternative would be to require all constant metadata columns to provide extractors, and define a default extractor that just goes to the map... but this seems like extra boilerplate for no real benefit? 

(I originally wanted to just embed the extractor in the attribute definition directly, but attribute/struct metadata cannot store code and IMO that's a wise design decision we shouldn't change)